charger/mennekes.go

```diff
@@ -0,0 +1,302 @@
+)
+
+func init() {
+	registry.Add("mennekes", NewMennekesFromConfig)
+}
+
+// NewMennekesFromConfig creates a new Mennekes ModbusTCP charger
+func NewMennekesFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		modbus.Settings `mapstructure:",squash"`
+		Timeout         time.Duration
+	}{
+		Settings: modbus.Settings{
+			Baudrate: 57600,
+			Comset:   "8N2",
+			ID:       50,
+		},
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewMennekes(cc.URI, cc.Device, cc.Comset, cc.Baudrate, modbus.ProtocolFromRTU(cc.RTU), cc.ID, cc.Timeout)
+}
+
+// NewMennekes creates Mennekes charger
+func NewMennekes(uri, device, comset string, baudrate int, proto modbus.Protocol, slaveID uint8, timeout time.Duration) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, device, comset, baudrate, proto, slaveID)
+	if err != nil {
+		return nil, err
+	}
+
+	if timeout > 0 {
+		conn.Timeout(timeout)
+	}
+
+	if !sponsor.IsAuthorized() {
+		return nil, api.ErrSponsorRequired
+	}
+
+	log := util.NewLogger("mennekes")
+	conn.Logger(log.TRACE)
+
+	wb := &Mennekes{
+		log:  log,
+		conn: conn,
+	}
+
+	// failsafe
+	go wb.heartbeat(time.Duration(mennekesHeartbeatInterval) * time.Second)
+
+	return wb, err
+}
+
+func (wb *Mennekes) heartbeat(timeout time.Duration) {
+	for range time.Tick(timeout) {
+		if _, err := wb.conn.WriteSingleRegister(mennekesRegHeartbeat, mennekesHeartbeatToken); err != nil {
+			wb.log.ERROR.Println("heartbeat:", err)
+		}
+	}
+}
+
+// Status implements the api.Charger interface
+func (wb *Mennekes) Status() (api.ChargeStatus, error) {
+	b, err := wb.conn.ReadHoldingRegisters(mennekesRegEvseState, 1)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	switch status := encoding.Uint16(b); status {
+	case 1:
+		return api.StatusA, nil
+
+	case 2, 3, 4:
+		return api.StatusB, nil
+
+	case 5:
+		return api.StatusC, nil
+
+	default:
+		return api.StatusNone, fmt.Errorf("invalid status: %d", status)
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Mennekes) Enabled() (bool, error) {
+	b, err := wb.conn.ReadHoldingRegisters(mennekesRegChargingReleaseEM, 1)
+	if err != nil {
+		return false, err
+	}
+	u := encoding.Uint16(b)
+	return u == mennekesAllowed, nil
+}
+
+// Enable implements the api.Charger interface
+func (wb *Mennekes) Enable(enable bool) error {
+	var u uint16
+	if enable {
+		u = 1
+	}
+	_, err := wb.conn.WriteSingleRegister(mennekesRegChargingReleaseEM, u)
+	return err
+}
+
+// MaxCurrent implements the api.Charger interface
+func (wb *Mennekes) MaxCurrent(current int64) error {
+	return wb.MaxCurrentMillis(float64(current))
+}
+
+var _ api.ChargerEx = (*Mennekes)(nil)
+
+// MaxCurrentMillis implements the api.ChargerEx interface
+func (wb *Mennekes) MaxCurrentMillis(current float64) error {
+	b := make([]byte, 4)
+	// TODO: Write in LSW first order
+	binary.BigEndian.PutUint32(b, math.Float32bits(float32(current)))
```

Comment:
@andig Wie mache ich das denn am geschicktesten?

Reply:
Reicht `LttleEndian`?

Reply:
Oder LE in Häppchen von 2 `uint16`?

Reply:
Nein, schon Big Endian Byte Order aber mit vertauschten Words. 
Also genau dass was enconding.xxxLswFirst() decodiert hier als Encoding/Put.

Notfalls mach ich es halt mit passendem manuellem Byte-Shifting.
Hatte nur gehofft dass das auch schon mal irgendwo implementiert ist.

Reply:
Mach 2x PutUint16 für loe und high Teil

Reply:
ping ;)